Added charts and development mode

alecello opened the pull request:
# Graphs
## Generic charts
* One pie chart for all contributions in the past year divided by repository
* One pie chart for the total SLOC count divided by repository
* One bar chart for contributions to all repositories in the past year divided by contributor
* One bar chart for contributions to all repositories in total time divided by contributor
## Per-repository charts
* One pie chart with the total lines of code divided by SLOC, comments and blanks (only if using cloc)
* One bar graph with contributions in the past year divided by contributor
* One bar graph with contributions in total time divided by contributor
> Note: per-repository bar graphs are generated only if there are at least two contributors

This should close #3
# Developer mode
Added one file config.py and two options:
* `devmode`: Enables local caching of API requests
* `keep_repos`: Enable long-term caching of cloned repositories
I tested this feature and it's entirely possible to work with no internet connection if both options are enabled. Should close #8

alecello committed:
Added plotting (not yet functional)
The graph has repeated colours and the legend is all over the place.

alecello committed:
Fixed legend position and bounding box, improved colors

Co-authored-by: e-caste-mbp <en.cas.dev@icloud.com>

alecello committed:
Refactored color sequence generation

alecello committed:
Added toggle to skip SLOC stats generation

alecello committed:
Added more graphs (still WIP)

alecello committed:
Improved bar graphs

alecello committed:
Fixed labeling in bar graphs

alecello committed:
Changed file and chart names

alecello committed:
Added SLOC stats

alecello committed:
Sort the charts and make pie charts start upright and go clockwise

alecello committed:
Added developer mode (local caching) and updated README.
Also changed output path to go in a dedicated sub-directory to minimize clutter

alecello committed:
Added configuration file config.py

alecello committed:
Changed defaults in config.py, added SLOC pie chart also for wc

e-caste commented:
The result is good. I consider #8 closed, but before closing #3, here are some suggestions:  

`matplotlib` allows to add multiple graphs to the same plot. This could be useful for 
- having an aggregation of the 3 repository-specific graphs into one single .svg file, instead of a directory of 3 .svg files
- having an aggregation of the 4 organisation-wide stats into a single .svg file, in addition to (not instead of) the separate files (they may be useful to show some data)

The pie charts without a number of what they represent don't carry much information. Could you add it?

A couple of example graphs should be added to the README imo to show how cool and good we are asd

e-caste commented on main.py:
```diff
@@ -157,21 +229,120 @@ def get_lines_stats(repos: list, use_cloc: bool) -> dict:
-def print_all_stats(commits_stats: dict, lines_stats: dict, contributors_stats: dict, use_cloc: bool):
+def generate_chart(data: dict, minimum: int, type: str, legend: str, title: str, path: str):
+    # Order data dictionary by size of elements
+    ls = sorted(data.items(), key=lambda x: int(x[1]))
```
One-liner version:
```suggestion
    ls = sorted(data.items(), key=lambda x: int(x[1]), reverse=True)
```

e-caste commented on main.py:
```diff
@@ -55,16 +73,36 @@ def get_anonymous_commits_stats(repos: list, header: dict) -> dict:
+
+            # If in devmode, cache the response in case it does not yet exist
+            if devmode:
+                with open(f'repo-stats/{repo}.anonymous.json', 'w') as f:
```
This kind of path concatenations doesn't work on Windows. Since we don't like it, but sometimes we use it, we have to support it. Use `os.path.join(dir, file)` for platform-agnostic code.

alecello replied:
Fixed with 1b3b965 and tested on Windows 10

alecello committed:
One-line dictionary sorting in graph generation

Thanks to @e-caste

Co-authored-by: e-caste <48513706+e-caste@users.noreply.github.com>

alecello committed:
Made paths platform independent

alecello committed:
Refactoring

alecello committed:
Added graphs to readme

alecello committed:
Fixed README

alecello committed:
Refactored graph generation code
This is the framework for implementing combined graphs

alecello committed:
Refactoring

alecello committed:
Added combined graph for general stats.
TODO: Reimplement all other stats
TODO: Make it such that pie and bar graphs are grouped together

e-caste commented on main.py:
```diff
@@ -17,6 +18,33 @@
+
+    title = 'Default graph title'
+    legend = 'Default graph legend'
+    kind = 'pie'
```
these variables may work, but it's usually better to provide them as default variables to the `__init__` method. E.g: `def func(a: int = 0, b: str = "string")`. Also, when a function starts having many parameters, it's more readable to put each one of them on a different line.

alecello replied:
Should be fixed with af9cefa. I did however use continuation lines with spacing not aligned to indentation level to keep them even with the first argument in the first line, which is supported but i get is a controversial topic among developers, so I'm not marking as solved yet.

alecello committed:
Refactor, add requirements.txt

alecello commented:
Commit af9cefa should also close #12

alecello committed:
Added more data to title and legends of graphs

alecello committed:
Re-added per-repository plots, fixed spacing in subplots

alecello committed:
Refactor graph generation system

alecello committed:
Sorted subplots by type

e-caste commented:
One last thing: the aggregate graphs for the organisation should be in a separate directory inside the output, so that they're immediately recognisable (e.g. WEEE-Open is a directory, peracotta.svg is an image file)

alecello committed:
Save general stats in a separate directory, capitalize organization name
